### PR TITLE
Specify docker directory instead of docker file for Flex custom runtime

### PR DESCRIPTION
**Pull request by nbashirbello**
Fix for issue #1304

### CODE REVIEW

File: .idea/compiler.xml [hunk 1]
```diff
@@ -1,7 +1,6 @@
 <?xml version="1.0" encoding="UTF-8"?>
 <project version="4">
   <component name="CompilerConfiguration">
-    <option name="DEFAULT_COMPILER" value="error-prone" />
```
Comment (etanshaul) [inline .idea/compiler.xml]: don't commit this; to prevent this from happening, install the error-prone compile plugin in IntelliJ and then make sure you set the java compiler (in IJ settings) to the error prone compiler

Comment (nbashirbello) [inline .idea/compiler.xml]: Thanks

**Comment by nbashirbello** [comment]
PR ready for review

File: google-cloud-tools-plugin/src/com/google/cloud/tools/intellij/appengine/cloud/CloudSdkAppEngineHelper.java [hunk 2]
```diff
@@ -148,7 +149,7 @@ public Project getProject() {
           return Optional.empty();
         }
         if (runtimeOptional.filter(runtime -> runtime == FlexibleRuntime.CUSTOM).isPresent()
-            && !Files.exists(Paths.get(deploymentConfiguration.getDockerFilePath()))) {
+            && !Files.exists(Paths.get(deploymentConfiguration.getDockerDirectoryPath() + DOCKERFILE))) {
```
Comment (etanshaul) [inline google-cloud-tools-plugin/src/com/google/cloud/tools/intellij/appengine/cloud/CloudSdkAppEngineHelper.java]: There should be a better way to do this: like using one of the file api's to check if a file exists within the directory. This avoids the concatenation and potential OS issues (may not work in Windows).

File: google-cloud-tools-plugin/src/com/google/cloud/tools/intellij/appengine/cloud/flexible/AppEngineFlexibleDeploymentEditor.java [hunk 3]
```diff
@@ -298,14 +299,14 @@ protected void resetEditorFrom(@NotNull AppEngineDeploymentConfiguration configu
...
-    dockerfileTextField.setText(configuration.getDockerFilePath());
+    dockerDirectoryTextField.setText(configuration.getDockerDirectoryPath());
     gcpProjectSelector.setText(configuration.getCloudProjectName());
     appYamlTextField.setVisible(configuration.isOverrideAppYaml()
         || modulesWithFlexFacetComboBox.getItemCount() == 0);
     archiveSelector.setText(configuration.getUserSpecifiedArtifactPath());
     appYamlOverrideCheckBox.setSelected(configuration.isOverrideAppYaml()
         || modulesWithFlexFacetComboBox.getItemCount() == 0);
-    dockerfileOverrideCheckBox.setSelected(configuration.isOverrideDockerfile()
+    dockerDirectoryOverrideCheckBox.setSelected(configuration.isOverrideDockerfile()
```
Comment (etanshaul) [inline google-cloud-tools-plugin/src/com/google/cloud/tools/intellij/appengine/cloud/flexible/AppEngineFlexibleDeploymentEditor.java]: should the method name `isOverrideDockerfile` in the configuration class also be updated?

File: google-cloud-tools-plugin/src/com/google/cloud/tools/intellij/appengine/cloud/flexible/AppEngineFlexibleDeploymentEditor.java [hunk 4]
```diff
@@ -374,18 +375,19 @@ private void validateConfiguration() throws ConfigurationException {
...
       if (isCustomRuntime()) {
-        if (StringUtils.isBlank(getDockerfilePath())) {
+        String dockerDirectoryPath = getDockerDirectoryPath();
+        if (StringUtils.isBlank(dockerDirectoryPath)) {
           throw new ConfigurationException(
-              GctBundle.message("appengine.flex.config.browse.dockerfile"));
+              GctBundle.message("appengine.flex.config.browse.docker.directory"));
         }
-        if (!isValidConfigurationFile(getDockerfilePath())) {
+        if (!isValidConfigurationFile(dockerDirectoryPath + DOCKERFILE)) {
```
Comment (etanshaul) [inline google-cloud-tools-plugin/src/com/google/cloud/tools/intellij/appengine/cloud/flexible/AppEngineFlexibleDeploymentEditor.java]: same comment here as above about concatenating to the path

File: google-cloud-tools-plugin/src/com/google/cloud/tools/intellij/appengine/cloud/flexible/AppEngineFlexibleStage.java [hunk 5]
```diff
@@ -72,7 +76,8 @@ public void stage(@NotNull Path stagingDirectory) {
             GctBundle.getString("appengine.deployment.error.staging.yaml"));
       }
       if (isCustomRuntime
-          && !Files.exists(Paths.get(deploymentConfiguration.getDockerFilePath()))) {
+          && !Files.exists(Paths.get(
+          deploymentConfiguration.getDockerDirectoryPath() + DOCKERFILE))) {
```
Comment (etanshaul) [inline google-cloud-tools-plugin/src/com/google/cloud/tools/intellij/appengine/cloud/flexible/AppEngineFlexibleStage.java]: same thing

File: google-cloud-tools-plugin/src/com/google/cloud/tools/intellij/appengine/facet/flexible/FlexibleFacetEditor.java [hunk 6]
```diff
@@ -186,9 +187,17 @@ public void apply() throws ConfigurationException {
...
-        throw new ConfigurationException(
-            GctBundle.getString("appengine.deployment.error.staging.dockerfile"));
+      if (isRuntimeCustom()) {
+        String dockerfileDir = dockerDirectory.getText();
+        if (dockerfileDir.isEmpty() || !Files.isDirectory(Paths.get(dockerfileDir))) {
+          throw new ConfigurationException(
+              GctBundle.getString("appengine.deployment.error.staging.docker.directory"));
+        }
+
+        if (!isValidConfigurationFile(dockerfileDir + DOCKERFILE)) {
```
Comment (etanshaul) [inline google-cloud-tools-plugin/src/com/google/cloud/tools/intellij/appengine/facet/flexible/FlexibleFacetEditor.java]: same

File: google-cloud-tools-plugin/src/com/google/cloud/tools/intellij/appengine/facet/flexible/FlexibleFacetEditor.java [hunk 7]
```diff
@@ -237,16 +246,25 @@ private void showWarnings() {
...
+    } else {
+      try {
+        if (isRuntimeCustom()) {
+          String dockerDir = dockerDirectory.getText();
+          if (dockerDir.isEmpty() || !Files.isDirectory(Paths.get(dockerDir))) {
+            errorMessage
+                .setText(
+                    GctBundle.getString("appengine.deployment.error.staging.docker.directory"));
+            showError = true;
+          } else if (!isValidConfigurationFile(dockerDir + DOCKERFILE)) {
```
Comment (etanshaul) [inline google-cloud-tools-plugin/src/com/google/cloud/tools/intellij/appengine/facet/flexible/FlexibleFacetEditor.java]: same

File: google-cloud-tools-plugin/src/com/google/cloud/tools/intellij/appengine/facet/flexible/FlexibleFacetEditor.java [hunk 8]
```diff
@@ -237,16 +246,25 @@ private void showWarnings() {
...
+            errorMessage
+                .setText(GctBundle.getString("appengine.deployment.error.staging.dockerfile"));
```
Comment (etanshaul) [inline google-cloud-tools-plugin/src/com/google/cloud/tools/intellij/appengine/facet/flexible/FlexibleFacetEditor.java]: is this the right name for the error message? why "error staging"?

File: google-cloud-tools-plugin/src/com/google/cloud/tools/intellij/appengine/facet/flexible/FlexibleFacetEditor.java [hunk 9]
```diff
@@ -287,19 +306,35 @@ public void onFacetInitialized(@NotNull Facet facet) {
...
       this.fileName = fileName;
       this.configFileGenerator = configFileGenerator;
-      this.filePicker = filePicker;
+      this.directoryPicker = directoryPicker;
       this.configurationValidator = configurationValidator;
     }
 
     @Override
     public void actionPerformed(ActionEvent event) {
+      boolean genAppYaml = "app.yaml".equals(fileName);
```
Comment (etanshaul) [inline google-cloud-tools-plugin/src/com/google/cloud/tools/intellij/appengine/facet/flexible/FlexibleFacetEditor.java]: this feels a bit off. this is supposed to be a generic generator yet you add app.yaml specific stuff here. I'm thinking this can be refactored to avoid this.

**Comment by etanshaul** [comment]
General observation:

for this change to actually be useful, don't we also have to update the Dockerfile template to pull in the contents of the directory? As it currently stands, only the war / jar artifact is added.

@patflynn @nbashirbello

**Comment by patflynn** [comment]
For custom use-cases we should expect the user to modify the Dockerfile so I think it's OK as is, but you're right that we could add the contents of the Docker dir.  The problem is that we don't want to add the full content of the staging directory, and that's where the docker build will happen.

I wonder what the spotify plugin does.

File: google-cloud-tools-plugin/resources/messages/CloudToolsBundle.properties [hunk 10]
```diff
@@ -90,7 +90,7 @@ appengine.flex.config.user.specified.artifact.title=Select a JAR or WAR file
 appengine.flex.config.user.specified.artifact.error=Browse to a JAR or WAR file.
 appengine.flex.config.custom.deploy.title=App Engine flexible environment Configuration
 appengine.flex.config.browse.app.yaml=Browse to an app.yaml file.
-appengine.flex.config.browse.dockerfile=Browse to a Dockerfile.
+appengine.flex.config.browse.docker.directory=Browse to a docker directory.
```
Comment (patflynn) [inline google-cloud-tools-plugin/resources/messages/CloudToolsBundle.properties]: I think you should capitalize docker

File: google-cloud-tools-plugin/resources/messages/CloudToolsBundle.properties [hunk 11]
```diff
@@ -122,8 +122,10 @@ appengine.stop.modules.version.canceled.message=Undeploy Canceled
 appengine.deployment.error.during.staging=Deployment failed due to staging failing with exit code: {0}.
 appengine.deployment.exception.during.staging=Deployment failed due to an exception while staging the project.
 appengine.deployment.error.staging.yaml=The specified app.yaml configuration file does not exist or is not a valid file.
-appengine.deployment.error.staging.dockerfile=The specified Dockerfile configuration file does not exist or is not a valid file.
-appengine.deployment.error.staging.gotosettings=Set a valid file in Module Settings or use another one.
+appengine.deployment.error.staging.dockerfile=There is no Dockerfile in specified directory or it is not a valid file.
```
Comment (patflynn) [inline google-cloud-tools-plugin/resources/messages/CloudToolsBundle.properties]: what is a 'valid file'?

File: google-cloud-tools-plugin/resources/messages/CloudToolsBundle.properties [hunk 12]
```diff
@@ -122,8 +122,10 @@ appengine.stop.modules.version.canceled.message=Undeploy Canceled
...
+appengine.deployment.error.staging.docker.directory=The specified docker directory does not exist or it is not a valid directory.
```
Comment (patflynn) [inline google-cloud-tools-plugin/resources/messages/CloudToolsBundle.properties]: same here.

File: google-cloud-tools-plugin/resources/messages/CloudToolsBundle.properties [hunk 13]
```diff
@@ -122,8 +122,10 @@ appengine.stop.modules.version.canceled.message=Undeploy Canceled
...
+appengine.deployment.error.staging.directory.gotosettings=Set a valid directory in Module Settings or use another one.
```
Comment (patflynn) [inline google-cloud-tools-plugin/resources/messages/CloudToolsBundle.properties]: I'd be tempted to use a key label that matched what it's used for instead of what it says.  Maybe appengine.deployment.error.staging.directory.invalid?

File: google-cloud-tools-plugin/resources/messages/CloudToolsBundle.properties [hunk 14]
```diff
@@ -122,8 +122,10 @@ appengine.stop.modules.version.canceled.message=Undeploy Canceled
...
+appengine.deployment.error.staging.file.gotosettings=Set a valid file in Module Settings or use another one.
```
Comment (patflynn) [inline google-cloud-tools-plugin/resources/messages/CloudToolsBundle.properties]: same here.

File: google-cloud-tools-plugin/src/com/google/cloud/tools/intellij/appengine/cloud/CloudSdkAppEngineHelper.java [hunk 15]
```diff
@@ -78,6 +78,7 @@
   private static final String REFRESH_TOKEN_LABEL = "refresh_token";
   private static final String GCLOUD_USER_TYPE_LABEL = "type";
   private static final String GCLOUD_USER_TYPE = "authorized_user";
+  private static final String DOCKERFILE = "/Dockerfile";
```
Comment (patflynn) [inline google-cloud-tools-plugin/src/com/google/cloud/tools/intellij/appengine/cloud/CloudSdkAppEngineHelper.java]: I'd be tempted to set this to 'Dockerfile', since '/Dockerfile' alone is a valid path to the root of the file system, which will never be what we want.

File: google-cloud-tools-plugin/src/com/google/cloud/tools/intellij/appengine/cloud/CloudSdkAppEngineHelper.java (hunk 2 above)
Comment (patflynn) [inline google-cloud-tools-plugin/src/com/google/cloud/tools/intellij/appengine/cloud/CloudSdkAppEngineHelper.java]: Yes, Path supports a way to build paths.

File: google-cloud-tools-plugin/src/com/google/cloud/tools/intellij/appengine/cloud/flexible/AppEngineFlexibleDeploymentEditor.form [hunk 16]
```diff
@@ -180,12 +180,12 @@
...
             </properties>
           </component>
-          <component id="9e96a" class="javax.swing.JLabel" binding="dockerfileLabel">
+          <component id="9e96a" class="javax.swing.JLabel" binding="dockerDirectoryLabel">
             <constraints>
               <grid row="2" column="0" row-span="1" col-span="1" vsize-policy="0" hsize-policy="0" anchor="8" fill="0" indent="0" use-parent-layout="false"/>
             </constraints>
             <properties>
-              <text value="Dockerfile:"/>
+              <text value="Docker Directory:"/>
```
Comment (patflynn) [inline google-cloud-tools-plugin/src/com/google/cloud/tools/intellij/appengine/cloud/flexible/AppEngineFlexibleDeploymentEditor.form]: this should come from the resource bundle.

File: google-cloud-tools-plugin/src/com/google/cloud/tools/intellij/appengine/facet/flexible/FlexibleFacetEditor.java [hunk 17]
```diff
@@ -186,9 +187,17 @@ public void apply() throws ConfigurationException {
...
 
     try {
-      if (isRuntimeCustom() && !isValidConfigurationFile(dockerfile.getText())) {
-        throw new ConfigurationException(
-            GctBundle.getString("appengine.deployment.error.staging.dockerfile"));
+      if (isRuntimeCustom()) {
+        String dockerfileDir = dockerDirectory.getText();
+        if (dockerfileDir.isEmpty() || !Files.isDirectory(Paths.get(dockerfileDir))) {
+          throw new ConfigurationException(
+              GctBundle.getString("appengine.deployment.error.staging.docker.directory"));
```
Comment (patflynn) [inline google-cloud-tools-plugin/src/com/google/cloud/tools/intellij/appengine/facet/flexible/FlexibleFacetEditor.java]: this error and the one below should include the actually configured path.

Comment (patflynn) [inline google-cloud-tools-plugin/src/com/google/cloud/tools/intellij/appengine/facet/flexible/FlexibleFacetEditor.java]: @etanshaul 
@nbashirbello 
I think this the kind of thing Nikolay suggested we not do in apply methods #1408

File: google-cloud-tools-plugin/src/com/google/cloud/tools/intellij/appengine/facet/flexible/FlexibleFacetEditor.java [hunk 18]
```diff
@@ -237,16 +246,25 @@ private void showWarnings() {
...
       errorMessage.setText(GctBundle.getString("appengine.deployment.error.staging.yaml"));
       showError = true;
-    }
-
-    try {
-      if (isRuntimeCustom() && !isValidConfigurationFile(dockerfile.getText())) {
-        errorMessage.setText(GctBundle.getString("appengine.deployment.error.staging.dockerfile"));
+    } else {
+      try {
+        if (isRuntimeCustom()) {
```
Comment (patflynn) [inline google-cloud-tools-plugin/src/com/google/cloud/tools/intellij/appengine/facet/flexible/FlexibleFacetEditor.java]: this looks exactly like the code in apply.

File: google-cloud-tools-plugin/src/com/google/cloud/tools/intellij/appengine/facet/flexible/FlexibleFacetEditor.java (hunk 9 above)
Comment (patflynn) [inline google-cloud-tools-plugin/src/com/google/cloud/tools/intellij/appengine/facet/flexible/FlexibleFacetEditor.java]: +1.

File: google-cloud-tools-plugin/resources/messages/CloudToolsBundle.properties (hunk 11 above)
Comment (etanshaul) [inline google-cloud-tools-plugin/resources/messages/CloudToolsBundle.properties]: see my UX doc for proposed updates. perhaps those changes can be anticipated in this PR.

https://docs.google.com/document/d/1rE4Ir9UnRXoLnRxxxqF63KmBL8xkl_W1KOwqUwdtM6Y

I specifically do not have the "valid file" part because we don't do anything intelligent with the Dockerfile yet.

File: google-cloud-tools-plugin/src/com/google/cloud/tools/intellij/appengine/facet/flexible/FlexibleFacetEditor.java (hunk 17 above)
Comment (etanshaul) [inline google-cloud-tools-plugin/src/com/google/cloud/tools/intellij/appengine/facet/flexible/FlexibleFacetEditor.java]: yea, in my UX update docs I change this so that the facet dialog can be saved even with errors.

I'd suggest to make the minimal set of changes here in this PR to accommodate updating to a Docker directory since once we agree on the UX updates there will be a separate PR to revamp this.

**Comment by patflynn** [comment]
sgtm

**Comment by etanshaul** [comment]
@patflynn regarding adding the other assets to the Dockerfile:

I agree that as a first step we can expect users with runtime: custom to add what is needed to the Dockerfile.  One way around way you said about not adding everything in the staging directory, would be to instead add the files individually from the specified docker directory (instead of just everything in there).

File: google-cloud-tools-plugin/src/com/google/cloud/tools/intellij/appengine/facet/flexible/FlexibleFacetEditor.java (hunk 8 above)
Comment (nbashirbello) [inline google-cloud-tools-plugin/src/com/google/cloud/tools/intellij/appengine/facet/flexible/FlexibleFacetEditor.java]: This is shared with the deployment ui

File: google-cloud-tools-plugin/src/com/google/cloud/tools/intellij/appengine/cloud/CloudSdkAppEngineHelper.java [hunk 19]
```diff
@@ -148,7 +150,7 @@ public Project getProject() {
           return Optional.empty();
         }
         if (runtimeOptional.filter(runtime -> runtime == FlexibleRuntime.CUSTOM).isPresent()
-            && !Files.exists(Paths.get(deploymentConfiguration.getDockerFilePath()))) {
+            && (!new File(deploymentConfiguration.getDockerDirectoryPath(), DOCKERFILE).exists())) {
```
Comment (etanshaul) [inline google-cloud-tools-plugin/src/com/google/cloud/tools/intellij/appengine/cloud/CloudSdkAppEngineHelper.java]: instead of exists() i'd suggest to use isFile() to make sure its not a directory.

But better yet, I think you can use the Paths api:

```
Path path = Paths.get(deploymentConfiguration.getDockerDirectoryPath(), DOCKERFILE));
..
.. && (!Files.isRegularFile(path))...
```

Same comment for all other cases doing this logic

File: google-cloud-tools-plugin/src/com/google/cloud/tools/intellij/appengine/cloud/CloudSdkAppEngineHelper.java [hunk 20]
```diff
@@ -78,6 +79,7 @@
   private static final String REFRESH_TOKEN_LABEL = "refresh_token";
   private static final String GCLOUD_USER_TYPE_LABEL = "type";
   private static final String GCLOUD_USER_TYPE = "authorized_user";
+  private static final String DOCKERFILE = "Dockerfile";
```
Comment (etanshaul) [inline google-cloud-tools-plugin/src/com/google/cloud/tools/intellij/appengine/cloud/CloudSdkAppEngineHelper.java]: DOCKERFILE_NAME ?

File: google-cloud-tools-plugin/src/com/google/cloud/tools/intellij/appengine/facet/flexible/FlexibleFacetEditor.java [hunk 21]
```diff
@@ -278,28 +300,46 @@ public void onFacetInitialized(@NotNull Facet facet) {
...
     // Used to refresh the warnings.
     private final Runnable configurationValidator;
 
     GenerateConfigActionListener(
         Project project,
         String fileName,
         Consumer<Path> configFileGenerator,
-        TextFieldWithBrowseButton filePicker,
+        TextFieldWithBrowseButton directoryPicker,
+        boolean isDirectory,
```
Comment (etanshaul) [inline google-cloud-tools-plugin/src/com/google/cloud/tools/intellij/appengine/facet/flexible/FlexibleFacetEditor.java]: I'm not really a fan of passing in a boolean like this to switch between the logic.

Is it possible to normalize the directory outside of this class? e.g. if its not a directory, to grab the parent directory outside of here and pass it into this generator?

File: google-cloud-tools-plugin/src/com/google/cloud/tools/intellij/appengine/facet/flexible/FlexibleFacetEditor.java [hunk 22]
```diff
@@ -237,16 +248,26 @@ private void showWarnings() {
...
       errorMessage.setText(GctBundle.getString("appengine.deployment.error.staging.yaml"));
       showError = true;
-    }
-
-    try {
-      if (isRuntimeCustom() && !isValidConfigurationFile(dockerfile.getText())) {
-        errorMessage.setText(GctBundle.getString("appengine.deployment.error.staging.dockerfile"));
+    } else {
+      try {
+        if (isRuntimeCustom()) {
```
Comment (etanshaul) [inline google-cloud-tools-plugin/src/com/google/cloud/tools/intellij/appengine/facet/flexible/FlexibleFacetEditor.java]: may change depending on if/how you handle @patflynn's comment here, but as is its a little hard to follow this code. Could it be refactored for clarity (extracting named methods where appropriate etc.)?

File: google-cloud-tools-plugin/src/com/google/cloud/tools/intellij/appengine/facet/flexible/FlexibleFacetEditor.java [hunk 23]
```diff
@@ -237,16 +248,26 @@ private void showWarnings() {
...
+          String dockerDirectoryTest = dockerDirectory.getText();
+          if (dockerDirectoryTest.isEmpty() || !Files.isDirectory(Paths.get(dockerDirectoryTest))) {
```
Comment (etanshaul) [inline google-cloud-tools-plugin/src/com/google/cloud/tools/intellij/appengine/facet/flexible/FlexibleFacetEditor.java]: dockerDirectoryTe**x**t? or dockerDirectoryName? also you could do something like  StringUtils.isEmpty for a null safe check too.

File: google-cloud-tools-plugin/src/com/google/cloud/tools/intellij/appengine/cloud/flexible/AppEngineFlexibleDeploymentEditor.java [hunk 24]
```diff
@@ -374,18 +375,19 @@ private void validateConfiguration() throws ConfigurationException {
...
           throw new ConfigurationException(
-              GctBundle.message("appengine.flex.config.browse.dockerfile"));
+              GctBundle.message("appengine.flex.config.browse.docker.directory"));
         }
-        if (!isValidConfigurationFile(getDockerfilePath())) {
+        if (!isValidConfigurationFile(Paths.get(dockerDirectoryPath, DOCKERFILE).toString())) {
           throw new ConfigurationException(
               GctBundle.getString("appengine.deployment.error.staging.dockerfile") + " "
-                  + GctBundle.getString("appengine.deployment.error.staging.gotosettings"));
+                  + GctBundle.getString("appengine.deployment.error.staging.directory.gotosettings"));
```
Comment (etanshaul) [inline google-cloud-tools-plugin/src/com/google/cloud/tools/intellij/appengine/cloud/flexible/AppEngineFlexibleDeploymentEditor.java]: This message doesn't mention a Docker direction, just a "directory". Shouldn't it indicate that its the Docker directory we are talking about here?

File: google-cloud-tools-plugin/src/com/google/cloud/tools/intellij/appengine/cloud/flexible/AppEngineFlexibleDeploymentEditor.java [hunk 25]
```diff
@@ -374,18 +375,19 @@ private void validateConfiguration() throws ConfigurationException {
     if (!isValidConfigurationFile(getAppYamlPath())) {
       throw new ConfigurationException(
           GctBundle.getString("appengine.deployment.error.staging.yaml") + " "
-              + GctBundle.getString("appengine.deployment.error.staging.gotosettings"));
+              + GctBundle.getString("appengine.deployment.error.staging.file.gotosettings"));
```
Comment (etanshaul) [inline google-cloud-tools-plugin/src/com/google/cloud/tools/intellij/appengine/cloud/flexible/AppEngineFlexibleDeploymentEditor.java]: Similar comment here as for the Docker directory messaging: this just mentions an "invalid file", not a bad app.yaml which is why the error is shown.

Aren't these messages shown at the bottom of the deployment config? If so they should mention what they are for.

File: google-cloud-tools-plugin/src/com/google/cloud/tools/intellij/appengine/facet/flexible/FlexibleFacetEditor.form [hunk 26]
```diff
@@ -65,7 +65,7 @@
               <grid row="2" column="0" row-span="1" col-span="1" vsize-policy="0" hsize-policy="1" anchor="8" fill="0" indent="0" use-parent-layout="false"/>
             </constraints>
             <properties>
-              <text value="Generate..."/>
+              <text value="Generate Dockerfile..."/>
```
Comment (etanshaul) [inline google-cloud-tools-plugin/src/com/google/cloud/tools/intellij/appengine/facet/flexible/FlexibleFacetEditor.form]: I know it was hardcoded here before..but this should also be in the message bundle.

File: google-cloud-tools-plugin/src/com/google/cloud/tools/intellij/appengine/cloud/flexible/AppEngineFlexibleDeploymentEditor.java (hunk 25 above)
Comment (nbashirbello) [inline google-cloud-tools-plugin/src/com/google/cloud/tools/intellij/appengine/cloud/flexible/AppEngineFlexibleDeploymentEditor.java]: Updated

File: google-cloud-tools-plugin/src/com/google/cloud/tools/intellij/appengine/facet/flexible/FlexibleFacetEditor.java (hunk 21 above)
Comment (nbashirbello) [inline google-cloud-tools-plugin/src/com/google/cloud/tools/intellij/appengine/facet/flexible/FlexibleFacetEditor.java]: We also need to know, after we generate the dockerfile of directory, if we are setting the directoryPicker to have the location of the file it just generated or its parent

File: google-cloud-tools-plugin/resources/messages/CloudToolsBundle.properties (hunk 11 above)
Comment (nbashirbello) [inline google-cloud-tools-plugin/resources/messages/CloudToolsBundle.properties]: Should we use regular file instead, because we are checking for that: There is no Dockerfile in specified directory or it is not a regular file.

Comment (etanshaul) [inline google-cloud-tools-plugin/resources/messages/CloudToolsBundle.properties]: I'm not sure what you mean by regular file? Not a directory (if so, I'd say that's implied and not worth the potentially confusing wording)?

File: google-cloud-tools-plugin/src/com/google/cloud/tools/intellij/appengine/facet/flexible/FlexibleFacetEditor.java (hunk 21 above)
Comment (etanshaul) [inline google-cloud-tools-plugin/src/com/google/cloud/tools/intellij/appengine/facet/flexible/FlexibleFacetEditor.java]: ok. Ideally, this would be refactored a somehow, but if it ends up not being cleaner, then at a minimum I'd suggest to document the behavior of this class then since it takes a little effort to decipher as is - what the outcome of setting `isDirectory` one way or the other etc.

File: google-cloud-tools-plugin/src/com/google/cloud/tools/intellij/appengine/facet/flexible/FlexibleFacetEditor.java [hunk 27]
```diff
@@ -321,20 +349,40 @@ public void actionPerformed(ActionEvent event) {
...
+            destinationFilePath.toString());
         configurationValidator.run();
       }
     }
   }
 
+  /**
+   * An object representing the outcome of a configuration validation check.
+   */
+  static class Result {
```
Comment (etanshaul) [inline google-cloud-tools-plugin/src/com/google/cloud/tools/intellij/appengine/facet/flexible/FlexibleFacetEditor.java]: ValidationResult ?

File: google-cloud-tools-plugin/src/com/google/cloud/tools/intellij/appengine/facet/flexible/FlexibleFacetEditor.java [hunk 28]
```diff
@@ -166,33 +171,23 @@ public JComponent createComponent() {
...
-
-    try {
-      if (isRuntimeCustom() && !isValidConfigurationFile(dockerfile.getText())) {
-        throw new ConfigurationException(
-            GctBundle.getString("appengine.deployment.error.staging.dockerfile"));
-      }
-    } catch (MalformedYamlFileException myf) {
-      throw new ConfigurationException(
-          GctBundle.getString("appengine.appyaml.malformed"));
+    Result result = validateConfiguration();
```
Comment (etanshaul) [inline google-cloud-tools-plugin/src/com/google/cloud/tools/intellij/appengine/facet/flexible/FlexibleFacetEditor.java]: why call validateConfiguration twice? showWarnings already calls it. I know that showWarnings is called from other places than apply(). You could do something like return the validationResult from showWarnings (and rename it to validateAndShowWarnings, or something).

**Comment by etanshaul** [comment]
@nbashirbello Aside from the last few comments I just made, things look good.

I noticed that there are some failing tests- its possible that if you haven't merged from master in a while then you are missing some of my fixes to the flaky tests. If thats the case, updating should fix it.

File: google-cloud-tools-plugin/src/com/google/cloud/tools/intellij/appengine/facet/flexible/FlexibleFacetEditor.java (hunk 21 above)
Comment (nbashirbello) [inline google-cloud-tools-plugin/src/com/google/cloud/tools/intellij/appengine/facet/flexible/FlexibleFacetEditor.java]: Done

File: google-cloud-tools-plugin/src/com/google/cloud/tools/intellij/appengine/facet/flexible/FlexibleFacetEditor.java [hunk 29]
```diff
@@ -278,28 +289,59 @@ public void onFacetInitialized(@NotNull Facet facet) {
...
     private final String fileName;
-    private final TextFieldWithBrowseButton filePicker;
+    private final TextFieldWithBrowseButton directoryPicker;
     private final Consumer<Path> configFileGenerator;
+    private final boolean isDirectory;
     // Used to refresh the warnings.
     private final Runnable configurationValidator;
 
+    /**
+     * Constructor
```
Comment (etanshaul) [inline google-cloud-tools-plugin/src/com/google/cloud/tools/intellij/appengine/facet/flexible/FlexibleFacetEditor.java]: I don't think this is proper javadoc /  google format. I'd add a sentence (and I think it has to end in a period).
